include/ic_material.h

```diff
@@ -0,0 +1,75 @@
+#pragma once
+
+#include <glm/glm.hpp>
+
+#include <map>
+#include <memory>
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace IC {
+    enum MaterialFlags {
+        None = 0,
+        Lit = 1 << 0,
+        Transparent = 1 << 1
+    };
+
+    enum class BindingType {
+        Uniform,
+        Texture
+    };
+
+    enum class ShaderDataType {
+        Float32,
+        String,
+        Vec2,
+        Vec3,
+        Vec4
+    };
+
+    struct ShaderBinding {
+        int index;
+        std::string name;
+        BindingType bindingType;
+        ShaderDataType dataType;
+    };
+
+    struct ShaderBindingValue {
+        ShaderBinding &binding;
+        void *value;
+        size_t size;
+    };
+
+    class MaterialTemplate {
+    public:
+        MaterialTemplate(){};
+        ~MaterialTemplate(){};
+
+        void AddBinding(int bindingIndex, std::string name, BindingType bindingType, ShaderDataType dataType);
+
+        std::map<int, ShaderBinding> &Bindings() { return _bindings; }
+
+        MaterialFlags flags;
+
+        std::string fragShaderData;
+        std::string vertShaderData;
+
+    private:
+        std::map<int, ShaderBinding> _bindings;
+    };
+
+    class MaterialInstance {
+    public:
+        MaterialInstance(MaterialTemplate &materialTemplate);
+        ~MaterialInstance(){};
+
+        std::map<int, std::shared_ptr<ShaderBindingValue>> BindingValues() { return _bindingValues; }
+        MaterialTemplate Template() { return _template; }
+        void SetBindingValue(int index, void *value, size_t size);
+
+    private:
+        MaterialTemplate &_template;
```

Comment:
Can we add `const` to this? I assume an instance would never want to modify the base template it's built from.

Reply:
this adds lots of issues, particularly with calling the function `Bindings()` from `_template`

Reply:
In theory those functions should then also be marked const, but if you don't want to deal with that it's fine